### PR TITLE
Adds a negative station trait: Vending products shortage.

### PR DESCRIPTION
**Pull request by Ghommie**
## About The Pull Request
Adds in a mean station trait that randomly lowers the availability of all vending products from vending machines on the station.

## Why It's Good For The Game
Kicking aside the simple concept of "The more the merrier (until it's a feature creep)" for a moment, vending (re)supplies are scarcely needed on a generic round basis. The addition of this new trait could lead to situations where more vending supplies are wanted/ordered once in a while.

Beside, this is a pretty evil negative trait notwithstanding its simplicity, unlike many other relatively innocuous ones.

## Changelog

:cl:
add: Added a 'Vending products shortage' station trait, that randomly lowers the availability of all vending products from vending machines on the station, with a 1/20 chance of the vending machine itself being tilted.
/:cl:

### CODE REVIEW

**Comment by carlarctg** [comment]
give carg techs bounty money for resupplying

**Comment by Ghommie** [comment]
> give carg techs bounty money for resupplying

Please, could you further articulate your interest-piquing idea instead of leaving me to perform mental gymnastic to interpret it and a possible implementation of such?

**Comment by AtomTheProphet** [comment]
rest in peace botany

**Comment by the-orange-cow** [comment]
I believe you can order vendor restocks through cargo, would the restock units suffer from this shortage?

> rest in peace botany

I suck at reading code, but would a vendor item have a chance of starting completely out of stock?  Botany HEAVILY relies on the megaseed vendor for seeds.  If it isn't already the case, there really should be at least one of each stock available shiftstart.

**Comment by ArcaneMusic** [comment]
> I suck at reading code, but would a vendor item have a chance of starting completely out of stock? Botany HEAVILY relies on the megaseed vendor for seeds. If it isn't already the case, there really should be at least one of each stock available shiftstart.

It could, but that's fine. A newly restocked vendor would have full stock. It's not like it's being billed as a positive modifer, anyway. It's your job to either request the supplies, or work around it if it inconvienances you.

> give carg techs bounty money for resupplying

See above.

**Comment by carlarctg** [comment]
> > give carg techs bounty money for resupplying
> 
> Please, could you further articulate your interest-piquing idea instead of leaving me to perform mental gymnastics to interpret it and its possible implementations?


make a bunch of machines start nearly-empty even on normal shifts
add a machine in cargo that tracks these drained vending machines and produces a resupplier from ORM materials
the vending machine produces a certification ticket (paper) when resupplied which can be fed to the cargo machine in exchange for bounty-level money

the 'cargo machine' can be the ordering console too, shrug

**Comment by san7890** [comment]
is a 2 weight too little?

**Comment by Ghommie** [comment]
What value would you suggest?

**Comment by Potato-Masher** [comment]
This is a good one, though I'd argue for some items just being outright out of stock. Station traits right now are a joke and don't really change up the round at all outside of "oh, the pdas make a funny sound" or "oh, the assistants have multicolored jumpsuits". Having the potential to change how people play the round on occasion is what station traits outta be for.

**Comment by Ghommie** [comment]
> This is a good one, though I'd argue for some items just being outright out of stock.

I appreciate the thought. If there would a case where we definitely want an item not to be out of stock,  a new variable could be added to limit it, I suppose.

File: code/modules/vending/_vending.dm [hunk 1]
```diff
@@ -225,14 +229,20 @@
...
+				onstation = FALSE
 			if(circuit)
 				circuit.onstation = onstation //sync up the circuit so the pricing schema is carried over if it's reconstructed.
+		else if(HAS_TRAIT(SSstation, STATION_TRAIT_VENDING_SHORTAGE))
+			for (var/datum/data/vending_product/product_record as anything in product_records + coin_records + hidden_records)
+				/**
+				 * in average, it should be 37.5% of the max amount, rounded up to the nearest int,
+				 * tho the max boundary can be as low/high as 50%/100%
+				 */
+				var/max_amount = rand(CEILING(product_record.amount/2, 1), product_record.amount)
```
Comment (Time-Green) [inline code/modules/vending/_vending.dm:244]:
```suggestion
				var/max_amount = rand(CEILING(product_record.amount * 0.5, 1), product_record.amount)
```
Multiplication is preferred when possible

File: code/datums/station_traits/negative_traits.dm [hunk 2]
```diff
@@ -17,6 +17,15 @@
 /datum/station_trait/distant_supply_lines/on_round_start()
 	SSeconomy.pack_price_modifier *= 1.2
 
+/datum/station_trait/vending_shortage
```
Comment (Time-Green) [inline code/datums/station_traits/negative_traits.dm:20]: Autodoc the type please

**Review by Time-Green — CHANGES_REQUESTED** [review]

**Comment by Time-Green** [comment]
Would be cool if it pre-crippled some assistants

**Pull request merged**